### PR TITLE
[asg-hevc] Validation tools for hevc_fei_sample were added

### PR DESCRIPTION
**Pull request by romka0017**
Commit includes:
1) Artificial stream generator (asg-hevc) for creation of the streams
and control structures.
2) hevc_fei_extractor for dumping information about CTUs and CUs from
bitstreams.
3) User manual.

Issue: n/a
Test: manually

### CODE REVIEW

File: tools/asg-hevc/include/block_structures.h [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef __BLOCK_STRUCTURES_H__
+#define __BLOCK_STRUCTURES_H__
+
+#include "mfxvideo.h"
+
+#if MFX_VERSION >= MFX_VERSION_NEXT
```
Comment (mgonchar) [inline tools/asg-hevc/include/block_structures.h]: replace VERSION_NEXT with 1027. hevc fei API added to 1.27

File: tools/asg-hevc/include/asglog.h [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef __ASGLOG_H__
+#define __ASGLOG_H__
+
+#include "mfxvideo.h"
+
+#if MFX_VERSION > 1024
```
Comment (dvrogozh) [inline tools/asg-hevc/include/asglog.h]: why 1024 if hevc appeared in 1027 as Miroslav suggested below?

File: tools/asg-hevc/include/frame_change_descriptor.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef __ASG_HEVC_FRAME_CHANGE_DESCRIPTOR_H__
+#define __ASG_HEVC_FRAME_CHANGE_DESCRIPTOR_H__
+
+#include "mfxvideo.h"
+
+#if MFX_VERSION >= MFX_VERSION_NEXT
```
Comment (dvrogozh) [inline tools/asg-hevc/include/frame_change_descriptor.h]: s/next/1027/?

File: tools/asg-hevc/include/frame_marker.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef __ASG_FRAME_MAKER_H__
+#define __ASG_FRAME_MAKER_H__
+
+#include "mfxvideo.h"
+
+#if MFX_VERSION >= MFX_VERSION_NEXT
```
Comment (dvrogozh) [inline tools/asg-hevc/include/frame_marker.h]: 1027?

File: tools/asg-hevc/include/frame_processor.h [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef __ASG_HEVC_FRAME_PROCESSOR_H__
+#define __ASG_HEVC_FRAME_PROCESSOR_H__
+
+#include "mfxvideo.h"
+
+#if MFX_VERSION >= MFX_VERSION_NEXT
```
Comment (dvrogozh) [inline tools/asg-hevc/include/frame_processor.h]: 1027?

**Comment by adydychk** [comment]
I want to mention that we do not have CI in  https://github.com/Intel-Media-SDK/tools at all.

File: tools/asg-hevc/include/asglog.h [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            m_LogFileOutput.close();
+        }
+    }
+
+private:
+    bool m_bVerbose = false;
+    std::ofstream m_LogFileOutput;
+};
+
+#endif // MFX_VERSION > 1024
```
Comment (mgonchar) [inline tools/asg-hevc/include/asglog.h]: please, correct the comment as well, just leave it as "MFX_VERSION"

**Comment by dvrogozh** [comment]
Folks, why this PR was merged without addressing comments? especially this one: "We need to decide whether we want to build and install tools always or we need to have ENABLE_TOOLS option =OFF by default."

**Comment by romka0017** [comment]
It will be discussed and fixed in the next commit.

**Comment by dvrogozh** [comment]
This should have been discussed and fixed in _this_ commit. That's the whole reason why we have a code review!

**Comment by romka0017** [comment]
Sorry, it is our mistake. It has not got effort on library or samples and will be done.

**Comment by dmitryermilov** [comment]
Roman said your concerns had been addressed.. Dmitry, do you suggest to revert?

**Comment by dvrogozh** [comment]
No, they were not. Major one is: add ENABLE_TOOLS=ON|OFF and there are other notes from me and Mirsolav regarding usage of MFX_VERSION_NEXT which I also see not addressed. I don't think we need to revert, but I would like us to be more careful in reviews. My request for changes is still there. And I would like to see commit addressing issues above pretty soon.

**Comment by romka0017** [comment]
Tool needs to use version next instead 1027.

**Comment by mgonchar** [comment]
@dvrogozh We decided to leave VERSION_NEXT in code, since this tool has some dependencies on structures which are still under API_VERSION_NEXT guard

**Comment by dvrogozh** [comment]
This (about version) should have been posted here before the merge.

Actually my major concern is miss of ENABLE_TOOLS. This is critical. Really critical. When you will add this?

**Comment by romka0017** [comment]
It will be added at next week

**Pull request merged**